<?xml version="1.0"?>
<launch>
  <arg name="mav_name"            default="slammer"/>
  <arg name="color"               default="Black"/>
  <arg name="x"                   default="2.5"/>
  <arg name="y"                   default="0"/>
  <arg name="z"                   default="0"/>
  <arg name="yaw"                 default="0"/>
  <arg name="debug"               default="false"/>
  <arg name="boat_name"               default="aruco"/>
  <arg name="show" default="true" />

  <!-- Start Simulator -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(find differential_flatness)/worlds/empty.world"/>
    <arg name="paused" value="true"/>
    <arg name="gui" value="true"/>
    <arg name="verbose" value="false"/>
    <arg name="debug" value="$(arg debug)"/>
  </include>

  <!-- Spawn Model in Gazebo -->
  <include file="$(find roscopter_sim)/launch/spawn_mav.launch">
    <arg name="mav_name"            value="$(arg mav_name)" />
    <arg name="model"               value="$(find differential_flatness)/agents/$(arg mav_name)/$(arg mav_name).urdf.xacro" />
    <arg name="x"                   value="$(arg x)"/>
    <arg name="y"                   value="$(arg y)"/>
    <arg name="z"                   value="$(arg z)"/>
    <arg name="yaw"                 value="$(arg yaw)"/>
    <arg name="color"               value="$(arg color)"/>
  </include>

  <!-- Controller -->
  <group ns="$(arg mav_name)">

    <!-- Load common parameters -->
    <rosparam command="load" file="$(find differential_flatness)/agents/$(arg mav_name)/$(arg mav_name).yaml"/>

    <!-- Is Flying Publisher -->
    <node pkg="rostopic" type="rostopic" name="is_flying_pub" args="pub is_flying std_msgs/Bool true"/>

    <!-- Status Publisher -->
    <node pkg="rostopic" type="rostopic" name="status_pub" args="pub -r 1 status rosflight_msgs/Status '{armed: true, failsafe: false, rc_override: false, offboard: true, error_code: 0, num_errors: 0, loop_time_us: 1}'"/>

    <!-- State Estimator -->
    <!-- <node pkg="roscopter" type="mekf" name="mekf" output="screen">
      <remap from="baro" to="baro/data"/>
      <remap from="sonar" to="sonar/data"/>
      <remap from="magnetometer" to="mag/data"/>
    </node> -->

    <!-- PID Position Controller -->
    <node name="controller" pkg="roscopter" type="controller" output="screen">
      <remap from="estimate" to="ground_truth/odometry/NED"/>
      <!-- <remap from="estimate" to="estimate"/> -->
    </node>

    <!-- LQR Controller -->
     <node name="LQR" pkg="differential_flatness" type="LQR.py" output="screen">
      <remap from="state" to="ground_truth/odometry/NED"/>
      <remap from="estimate" to="estimate"/>
    </node> 

    <!-- inverse -->
    <node name="inverse" pkg="differential_flatness" type="inverse.py" output="screen">
      <remap from="state" to="ground_truth/odometry/NED"/>
      <remap from="feed_forward_control" to="u_raw"/>
       <!--<remap from="command" to="diff_flat_cmd"/> -->
    </node>

    <!-- Attitude_control -->
    <node name="attitude_control" pkg="differential_flatness" type="attitude_control.py" output="screen">
      <remap from="state" to="ground_truth/odometry/NED"/>
    </node>

    <!-- Trajectory Manager -->
    <node name="trajectory" pkg="differential_flatness" type="traj.py" output="screen">
      <remap from="waypoint" to="high_level_command"/>
      <remap from="state" to="ground_truth/odometry/NED"/>
      <!-- <remap from="state" to="estimate"/> -->
    </node>

    <!-- plot states -->
    <node pkg="differential_flatness" type="states_plotter.py" name="states_plotter" output="screen">
      <param name="time_window" value="10.0"/>
    </node> 

    <!-- My estimator -->
    <!-- <node pkg="differential_flatness" type="differential_flatness2D.py" name="estimator_slam"/> -->

  </group>


</launch>
